design/one-pager-default-resource-class.md

```diff
@@ -88,66 +116,237 @@ spec:
+ability to select a *generic* "class of service" (i.e. `mysql-large`).
 
 ```yaml
-apiVersion: core.crossplane.io/v1alpha1
-kind: ResourceClass
+apiVersion: database.crossplane.io/v1alpha1
+kind: MySQLInstance
 metadata:
-  name: cloud-nosql
-  namespace: crossplane-system
-parameters:
-  ...
-defaultForClaimKinds:
-- nosqlinstance.database.crossplane.io
-- dynamodbinstance.storage.crossplane.io
-provisioner: dynamodbinstance.database.aws.crossplane.io/v1alpha1
-providerRef:
-  name: aws-provider
-reclaimPolicy: Delete
+  name: mysql-claim
+spec:
+  classRef:
+    kind: RDSInstanceClass
+    apiVersion: database.aws.crossplane.io/v1alpha1
+    name: standard-mysql
+    namespace: crossplane-system
+  writeConnectionSecretToRef:
+    name: rdsmysql
+  engineVersion: "5.6"
 ```
 
-And a resource claim that defaulted to the resource class above could look like this in the general case:
+To continue to provide the same level of portability for claims that was
+originally present, two enhancements to the `Policy` model can be made:
+
+1. Allow for multiple policies per namespace.
+1. Denote the default policy via label.
+
+### Multiple Policies Per Namespace
+
+Currently, default class controllers, which act on claims that omit a
+`classRef`, will fail to reconcile if multiple `Policies` exist within the
+claim's namespace. For example, if two `MySQLInstancePolicies` are created in
+namespace `crossplane-system`, then a `MySQLInstance` claim created in
+`crossplane-system` that omits a `classRef` will not be assigned either one of
+the classes referenced by the two `MySQLInstancePolicy` objects respectively.
+
+In order to reintroduce the ability to select a class of service in claim, the
+following steps can be taken:
+
+1. Update all `Policy` types to include a `classRef` field instead of a
+   `defaultClassRef` field. This would require updating the `crossplane-runtime`
+   embeddable struct
+   [`Policy`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L93),
+   then bumping the dependency in core Crossplane.
+1. Remove the `classRef` field from all claims entirely. This would require the
+   same steps as above but with
+   [`ResourceClaimSpec`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L49).
+1. Add a `providerRef` field to all claims using the same method detailed above.
+   This would require updating the same steps as above also with
+   [`ResourceClaimSpec`](https://github.com/crossplaneio/crossplane-runtime/blob/e4d61ee2805af680baf16fc2a1d8f79538d0f9bb/apis/core/v1alpha1/resource.go#L49).
+1. Alter all claim controllers (which live in provider stacks) to only reconcile
+   claims that have a `policyRef` to a `Policy` that has a `classRef` of that
+   kind's strongly-typed class. For instance, the `RDSInstance` claim controller
```

Comment:
What are the performance considerations to this approach, if any? I am having trouble imagining an efficient way to do this lookup without having every controller inspect every claim and its policy.

Reply:
To be totally honest, I am not quite sure of potential performance implications. However, I imagine a simple object reference lookup would be relatively inexpensive. This would obviously grow with number of claim reconcilers introduced though.

Reply:
As @hasheddan implies, we instantiate a resource claim controller for each `(resource claim, managed resource)` tuple. So there's a resource claim controller for `(MySQLInstance, RDSInstance)`, `(KubernetesCluster, GKECluster)`, etc. This allows resource claim controllers to live in cloud provider stacks, making it easy for new cloud providers to add managed resources that satisfy the set of portable resource claims defined by Crossplane core.

Before we introduced strongly typed resource classes the generic `ResourceClass` kind contained a `.provisioner` string that would specify the kind of managed resource that should be used to satisfy any claim referencing the class. This field read like `provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1`. This meant that the resource claim controllers determined which claims they should operate on by:

1. Watching for changes to all claims of the kind they were concerned with.
1. Using a watch predicate function to filter out claims that did not reference a `ResourceClass` whose `provisioner` mapped to the managed resource kind they were concerned with.

The predicate function mentioned in step 2 of the above process used a `controller-runtime` `client.Client` to `Get` the referenced `ResourceClass`. So every time a resource claim changed in the Kubernetes API, one controller for each managed resource kind that could possibly satisfy that claim would `Get` its `ResourceClass` to check whether they should. This sounds pretty bad, but fortunately the default `client.Client` instance is backed by [a cache](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/cache) of API objects so we were not likely to make an actual API call for each controller. We managed to cut out this `client.Client` when we switched to strongly typed resource classes, instead using the [`HasClassReferenceKind` predicate](https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#HasClassReferenceKind).

I imagine we'd return to a similar predicate under this proposal, only we'd `Get` and inspect the relevant policy instead. For example the `(MySQLInstance, RDSInstance)` resource claim controller would:

1. Watch for changes to all `MySQLInstance` resource claims.
1. Use a watch predicate to filter out claims that did not reference a `MySQLInstancePolicy` whose `classRef` was of `kind: RDSInstanceClass`. This would involve a `Get`, but frequently hit the cache as described above.